model/vm_host.rb

```diff
@@ -243,16 +244,83 @@ def reset
+    devices.map do |device_name|
+      result = ssh_session.exec!("sudo smartctl -j -H /dev/#{device_name} | jq .smart_status.passed").strip == "true"
+      result.tap do |passed|
+        Clog.emit("Device #{device_name} failed smartctl check.") unless passed
+      end
+    end.all?(true)
+  end
+
+  def check_storage_nvme(ssh_session, devices)
+    devices.map do |device_name|
+      result = ssh_session.exec!("sudo nvme smart-log /dev/#{device_name} | grep \"critical_warning\" | awk '{print $3}'").strip == "0"
+      result.tap do |passed|
+        Clog.emit("Device #{device_name} failed nvme smart-log check.") unless passed
+      end
+    end.all?(true)
+  end
+
+  def check_storage_read_write(ssh_session, devices)
+    lsblk_json_info = ssh_session.exec!("lsblk --json")
+    devices_with_mount_points = devices.map { |device| SystemParser.get_device_mount_points_from_lsblk_json(lsblk_json_info, device) }
+
+    all_mount_points = []
+    devices_with_mount_points.each do |device_mount_points|
+      device_mount_points.each do |device_name, mount_points|
+        mount_points.each do |mount_point|
+          all_mount_points.append(mount_point)
+        end
+      end
+    end
+
+    all_mount_points.uniq.all? do |mount_point|
+      file_name = (mount_point == "/") ? "/test-file" : "#{mount_point}/test-file"
+      write_status = ssh_session.exec!("sudo bash -c \"head -c 1M </dev/zero > #{file_name}\"").exitstatus == 0
+      hash_status = ssh_session.exec!("sha256sum #{file_name}").strip == "30e14955ebf1352266dc2ff8067e68104607e750abb9d3b36582b8af909fcb58  #{file_name}"
+      delete_status = ssh_session.exec!("sudo rm #{file_name}").exitstatus == 0
+
+      if !(write_status && hash_status && delete_status)
+        Clog.emit("failed to perform read/write on mountpoint #{mount_point}")
+      end
+      write_status && hash_status && delete_status
+    end
+  end
+
+  def check_storage_kernel_logs(ssh_session, devices)
+    kernel_logs = ssh_session.exec!("journalctl -kS -1min --no-pager")
+    return false unless kernel_logs.exitstatus == 0
+
+    error_count = kernel_logs.scan(/Buffer I\/O error on dev (\w+)/).tally
+    Clog.emit("found error on kernel logs. devices with error_count: #{error_count}") unless error_count.empty?
+    error_count.empty?
+  end
+
   def init_health_monitor_session
     {
       ssh_session: sshable.start_fresh_session
     }
   end
 
+  def disk_devices
+    storage_devices.each { |sd| sd.set_underlying_unix_devices if sd.unix_device_list.empty? }
+    storage_devices.flat_map { |sd| sd.unix_device_list.split(",") }
+  end
+
+  def perform_health_checks(ssh_session)
+    checks = [
+        -> { check_storage_smartctl(ssh_session, disk_devices) },
+        -> { check_storage_nvme(ssh_session, disk_devices) },
+        -> { check_storage_read_write(ssh_session, disk_devices) },
+        -> { check_storage_kernel_logs(ssh_session, disk_devices) }
+      ]
+      checks.all?(&:call)
```

Comment:
```suggestion
    check_storage_smartctl(ssh_session, disk_devices) &&
      check_storage_nvme(ssh_session, disk_devices) &&
      check_storage_read_write(ssh_session, disk_devices) &&
      check_storage_kernel_logs(ssh_session, disk_devices)
```

Best to avoid unnecessary indirection.